test/FastTests/Client/QueriesWithCustomFunctions.cs

```diff
@@ -1668,6 +1668,112 @@ select new
             }
         }
 
+        [Fact]
+        public void Custom_Function_First_And_FirstOrDefault_Support()
+        {
+            using (var store = GetDocumentStore())
+            {
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new Detail { Number = 1 }, "details/1");
+                    session.Store(new Detail { Number = 2 }, "details/2");
+                    session.Store(new Detail { Number = 3 }, "details/3");
+                    session.Store(new Detail { Number = 4 }, "details/4");
+
+                    session.Store(new User { Name = "Jerry", LastName = "Garcia", DetailIds = new List<string> { "details/1", "details/2" }}, "users/1");
+                    session.Store(new User { Name = "Bob", LastName = "Weir", DetailIds = new List<string> { "details/3", "details/4" } }, "users/2");
+
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var query = from u in session.Query<User>()
+                                let details = RavenQuery.Load<Detail>(u.DetailIds)
+                                select new 
+                                {
+                                    Name = u.Name,
+                                    First = details.First(x=>x.Number > 1).Number,
+                                    FirstOrDefault = details.FirstOrDefault(x => x.Number < 3)
+                                };
+
+                    Assert.Equal("FROM Users as u LOAD u.DetailIds as details[] " +
+                                 "SELECT { Name : u.Name, First : details.filter(function(x){return x.Number>1;})[0].Number, " +
```

Comment:
Filter iterates whole collection and this gives array bounds error if not found? Find would have correct semantics if supported https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find .